File: apps/user/src/dao/SuggestionsDao.ts
'use strict'

import {Db} from "mongodb";

const {MongoClient} = require('mongodb');
const client = new MongoClient(process.env.MONGO_URI);

import {SlimUser} from "../model/user/SlimUser";
import {User} from "../model/user/User";

const UserDocument = require('./document/UserDocument.ts')

const MIN_SHARED_CONNECTIONS = 2

class SuggestionsDao {
    private db: Db;
    private collectionName: string;
    private userCollectionName: string;

    constructor() {
        this.db = client.db('nevvi')
        this.collectionName = 'connection_suggestions'
        this.userCollectionName = 'users'
    }

    async findPossibleSuggestions(userId: string): Promise<User[]> {
        const pipeline: any = [
            {
                $match: {
                    _id: userId,
                },
            },
            {
                $lookup: {
                    from: "connections",
                    localField: "_id",
                    foreignField: "userId",
                    as: "connections",
                },
            },
            {
                $lookup: {
                    from: "connection_suggestions",
                    localField: "_id",
                    foreignField: "_id",
                    as: "suggestions",
                }
            },
            {
                $project: {
                    _id: 1,
                    connectionIds: "$connections.connectedUserId",
                    blockedUsers: 1,
                    ignoredSuggestions: {
                        "$ifNull": [{"$arrayElemAt": [ "$suggestions.ignored", 0 ]}, []]
                    }
                },
            },
            {
                $graphLookup: {
                    from: "connections",
                    startWith: "$_id",
                    connectFromField: "connectedUserId",
                    connectToField: "userId",
                    as: "connectionHierarchy",
                    maxDepth: 1,
                },
            },
            {
                $unwind: {
                    path: "$connectionHierarchy",
                },
            },
            {
                $project: {
                    connectionUserId: "$connectionHierarchy.userId",
                    suggestedUserId: "$connectionHierarchy.connectedUserId",
                    connectionIds: 1,
                    blockedUsers: 1,
                    ignoredSuggestions: 1
                },
            },
            {
                $addFields: {
                    connected: {
                        $in: [
                            "$suggestedUserId",
                            "$connectionIds",
                        ],
                    },
                    blocked: {
                        $in: [
                            "$suggestedUserId",
                            "$blockedUsers",
                        ],
                    },
                    ignored: {
                        $in: [
                            "$suggestedUserId",
                            "$ignoredSuggestions",
                        ],
                    }
                },
            },
            {
                $match: {
                    $and: [
                        {
                            connectionUserId: {
                                $ne: userId,
                            },
                        },
                        {
                            suggestedUserId: {
                                $ne: userId,
                            },
                        },
                        {
                            connected: false,
                        },
                        {
                            blocked: false,
                        },
                        {
                            ignored: false
                        }
                    ],
                },
            },
            {
                $group: {
                    _id: "$suggestedUserId",
                    count: {
                        $sum: 1,
                    },
                },
            },
            {
                $match: {
                    "count": {
                        $gte: MIN_SHARED_CONNECTIONS
                    }
                }
            },
            {
                $sort: {
                    count: -1,
                },
            },
            {
                $limit: 20,
            },
            {
                $lookup: {
                    from: "users",
                    localField: "_id",
                    foreignField: "_id",
                    as: "suggestedUser",
                },
            },
        ]

        const suggestions = await this.db.collection(this.userCollectionName)
            .aggregate(pipeline)
            .toArray()

        return suggestions.filter(suggestion => {
            return suggestion["suggestedUser"] && suggestion["suggestedUser"].length === 1
        }).map(suggestion => {
            const user = new User(suggestion["suggestedUser"][0])
            user.id = suggestion._id
            return user
        })
    }

    async updateSuggestions(userId: string, suggestions: string[]) {
        await this.db.collection(this.collectionName).updateOne(
            {_id: userId},
            {$set: {"suggestions": suggestions}},
            {upsert: true}
        )
    }

    async getSuggestions(userId: string): Promise<SlimUser[]> {
        const pipeline: any = [
            {
                '$match': {
                    '_id': userId
                }
            }, {
                '$lookup': {
                    'from': 'connections',
                    'localField': '_id',
                    'foreignField': 'userId',
                    'as': 'connections'
                }
            }, {
                '$project': {
                    'suggestions': {
                        '$filter': {
                            'input': '$suggestions',
                            'as': 'suggestion',
                            'cond': {
                                '$not': {
                                    '$in': [
                                        '$$suggestion', '$connections.connectedUserId'
                                    ]
                                }
                            }
                        }
                    }
                }
            }, {
                '$unwind': {
                    'path': '$suggestions',
                    'preserveNullAndEmptyArrays': false
                }
            }, {
                '$lookup': {
                    'from': 'users',
                    'localField': 'suggestions',
                    'foreignField': '_id',
                    'as': 'suggestedUser'
                }
            }
        ]

        const results = await this.db.collection(this.collectionName)
            .aggregate(pipeline)
            .toArray()

        // Return the user that we mapped over so that we don't need to grab all that info again 1 by 1...
        return results
            .filter(i => i["suggestedUser"] && i["suggestedUser"].length === 1)
            .map(i => {
                const user = new SlimUser(new UserDocument(i["suggestedUser"][0]))
                user.id = i["suggestedUser"][0]._id
                return user
            })
    }

    async removeSuggestion(userId: string, suggestedUserId: string): Promise<boolean> {
        const result = await this.db.collection(this.collectionName)
            .updateOne(
                {_id: userId},
                {$pull: {suggestions: suggestedUserId}}
            )

        return result.modifiedCount === 1
    }

    async ignoreSuggestion(userId: string, suggestedUserId: string): Promise<boolean> {
        const update = {
            $pull: {suggestions: suggestedUserId},
            $push: {ignored: suggestedUserId}
        }

        const result = await this.db.collection(this.collectionName)
            .updateOne({_id: userId}, update)

        return result.modifiedCount === 1
    }
}

export {SuggestionsDao}

File: apps/user/src/dao/suggestions_dao.py
import os
from typing import List, Dict, Any

import pymongo
from pymongo.synchronous.collection import Collection

from src.model.document import SuggestedUser

MIN_SHARED_CONNECTIONS = 2


class SuggestionsDao:
    def __init__(self):
        client = pymongo.MongoClient(os.environ.get("MONGO_URI"))
        self.suggestion_collection = client.get_database("nevvi").get_collection("connection_suggestions")
        self.user_collection = client.get_database("nevvi").get_collection("users")

    def find_possible_suggestions(self, user_id: str) -> List[SuggestedUser]:
        pipeline: List[Dict[str, Any]] = [
            {
                "$match": {
                    "_id": user_id,
                },
            },
            {
                "$lookup": {
                    "from": "connections",
                    "localField": "_id",
                    "foreignField": "userId",
                    "as": "connections",
                },
            },
            {
                "$lookup": {
                    "from": "connection_suggestions",
                    "localField": "_id",
                    "foreignField": "_id",
                    "as": "suggestions",
                }
            },
            {
                "$project": {
                    "_id": 1,
                    "connectionIds": "$connections.connectedUserId",
                    "blockedUsers": 1,
                    "ignoredSuggestions": {
                        "$ifNull": [{"$arrayElemAt": ["$suggestions.ignored", 0]}, []]
                    }
                },
            },
            {
                "$graphLookup": {
                    "from": "connections",
                    "startWith": "$_id",
                    "connectFromField": "connectedUserId",
                    "connectToField": "userId",
                    "as": "connectionHierarchy",
                    "maxDepth": 1,
                },
            },
            {
                "$unwind": {
                    "path": "$connectionHierarchy",
                },
            },
            {
                "$project": {
                    "connectionUserId": "$connectionHierarchy.userId",
                    "suggestedUserId": "$connectionHierarchy.connectedUserId",
                    "connectionIds": 1,
                    "blockedUsers": 1,
                    "ignoredSuggestions": 1
                },
            },
            {
                "$addFields": {
                    "connected": {
                        "$in": [
                            "$suggestedUserId",
                            "$connectionIds",
                        ],
                    },
                    "blocked": {
                        "$in": [
                            "$suggestedUserId",
                            "$blockedUsers",
                        ],
                    },
                    "ignored": {
                        "$in": [
                            "$suggestedUserId",
                            "$ignoredSuggestions",
                        ],
                    }
                },
            },
            {
                "$match": {
                    "$and": [
                        {
                            "connectionUserId": {
                                "$ne": user_id,
                            },
                        },
                        {
                            "suggestedUserId": {
                                "$ne": user_id,
                            },
                        },
                        {
                            "connected": False,
                        },
                        {
                            "blocked": False,
                        },
                        {
                            "ignored": False
                        }
                    ],
                },
            },
            {
                "$group": {
                    "_id": "$suggestedUserId",
                    "count": {
                        "$sum": 1,
                    },
                },
            },
            {
                "$match": {
                    "count": {
                        "$gte": MIN_SHARED_CONNECTIONS
                    }
                }
            },
            {
                "$sort": {
                    "count": -1,
                },
            },
            {
                "$limit": 20,
            },
            {
                "$lookup": {
                    "from": "users",
                    "localField": "_id",
                    "foreignField": "_id",
                    "as": "suggestedUser",
                },
            },
        ]

        return self.__execute_and_transform(pipeline, self.user_collection)

    def get_suggestions(self, user_id: str) -> List[SuggestedUser]:
        pipeline: List[Dict[str, Any]] = [
            {
                '$match': {
                    '_id': user_id
                }
            },
            {
                '$lookup': {
                    'from': 'connections',
                    'localField': '_id',
                    'foreignField': 'userId',
                    'as': 'connections'
                }
            },
            {
                '$project': {
                    'suggestions': {
                        '$filter': {
                            'input': '$suggestions',
                            'as': 'suggestion',
                            'cond': {
                                '$not': {
                                    '$in': [
                                        '$$suggestion', '$connections.connectedUserId'
                                    ]
                                }
                            }
                        }
                    }
                }
            },
            {
                '$unwind': {
                    'path': '$suggestions',
                    'preserveNullAndEmptyArrays': False
                }
            },
            {
                '$lookup': {
                    'from': 'users',
                    'localField': 'suggestions',
                    'foreignField': '_id',
                    'as': 'suggestedUser'
                }
            }
        ]

        return self.__execute_and_transform(pipeline, self.suggestion_collection)

    def update_suggestions(self, user_id: str, suggestions: List[str]):
        self.suggestion_collection.update_one(
            filter={"_id": user_id},
            update={"$set": {"suggestions": suggestions}},
            upsert=True
        )

    def remove_suggestions(self, user_id: str, suggested_user_id: str) -> bool:
        res = self.suggestion_collection.update_one(
            filter={"_id": user_id},
            update={"$pull": {"suggestions": suggested_user_id}},
        )

        return res.modified_count == 1

    def ignore_suggestions(self, user_id: str, suggested_user_id: str) -> bool:
        res = self.suggestion_collection.update_one(
            filter={"_id": user_id},
            update={
                "$pull": {"suggestions": suggested_user_id},
                "$push": {"ignored": suggested_user_id},
            },
        )

        return res.modified_count == 1

    @staticmethod
    def __execute_and_transform(pipeline: List[Dict[str, Any]], collection: Collection) -> List[SuggestedUser]:
        results = list(collection.aggregate(pipeline))
        suggested_users = [s.get("suggestedUser")[0] for s in results if len(s.get("suggestedUser", [])) == 1]

        def transform(result: Dict[str, Any]) -> SuggestedUser:
            return SuggestedUser(
                id=result.get("_id"),
                firstName=result.get("firstName"),
                lastName=result.get("lastName"),
                bio=result.get("bio"),
                profileImage=result.get("profileImage"),
                blockedUsers=result.get("blockedUsers")
            )

        return list(map(transform, suggested_users))
